lib/storage/src/content_manager/toc/collection_meta_ops.rs

```diff
@@ -392,28 +393,46 @@ impl TableOfContent {
                 )?;
                 collection.finish_shard_transfer(transfer).await?;
             }
-            ShardTransferOperations::SnapshotRecovered(transfer) => {
+            ShardTransferOperations::SnapshotRecovered(transfer)
+            | ShardTransferOperations::RecoveryToPartial(transfer) => {
                 // Validate transfer exists to prevent double handling
                 transfer::helpers::validate_transfer_exists(
                     &transfer,
                     &collection.state().await.transfers,
                 )?;
 
-                // Set shard state from `PartialSnapshot` to `Partial`
-                // TODO(1.9): get into Partial state from PartialSnapshot or Recovery
-                let operation = SetShardReplicaState {
-                    collection_name: collection_id,
-                    shard_id: transfer.shard_id,
-                    peer_id: transfer.to,
-                    state: ReplicaState::Partial,
-                    from_state: Some(ReplicaState::PartialSnapshot),
-                };
                 log::debug!(
-                    "Set shard replica state from {:?} to {:?} after snapshot recovery",
+                    "Set shard replica state from {:?} or {:?} to {:?} after snapshot recovery",
                     ReplicaState::PartialSnapshot,
+                    ReplicaState::Recovery,
                     ReplicaState::Partial,
                 );
-                self.set_shard_replica_state(operation).await?;
+
+                // Set shard state from `PartialSnapshot` or `Recovery` to `Partial`
+                self.get_collection(&collection_id)
+                    .await?
+                    .set_shard_replica_state_impl(
+                        transfer.shard_id,
+                        transfer.to,
+                        ReplicaState::Partial,
+                        |current_state| {
+                            match current_state {
+                                Some(ReplicaState::PartialSnapshot) => Ok(()),
+                                Some(ReplicaState::Recovery) => Ok(()),
+                                _ => {
+                                    Err(CollectionError::bad_input(format!(
+                                        "Replica {} of {collection_id}:{} has unexpected {current_state:?} \
+                                         (expected {:?} or {:?})",
+                                        transfer.to,
+                                        transfer.shard_id,
+                                        ReplicaState::PartialSnapshot,
+                                        ReplicaState::Recovery,
+                                    )))
+                                }
+                            }
+                        }
+                    )
+                    .await?;
```

Comment:
What about keeping this the same as what we had before, but requesting the current state first?

So:
1. get current replica set local shard state
2. assert it is PartialSnapshot/Recovery
3. call `set_shard_replica_state` with that `from` state

It looks like that would keep things simpler.

In 1.9 or 1.10 we likely remove `PartialSnapshot` in which case it'll only change from `Recovery` to `Partial` again.

Reply:
Done.